### PR TITLE
morphotagger/pymorphy_dev

### PR DESCRIPTION
**Pull request by AlexeySorokin**
Adds lemmatization to russian morphological tagger.

### CODE REVIEW

File: docs/components/morphotagger.rst [hunk 1]
```diff
@@ -82,8 +82,80 @@ You may also pass the tokenized sentences instead of raw ones:
     for parse in model(sentences):
         print(parse)
 
+If your data is large, you can call
+:func:`~deeppavlov.models.morpho_tagger.common.call_model` function, which will additionally
```
Comment (yoptar) [inline docs/components/morphotagger.rst]: I don't think this function should be there. We can just advise users to divide larger datasets into batches for inference:
```python3
from deeppavlov import build_model, configs
model = build_model(configs.morpho_tagger.UD2_0.morpho_ru_syntagrus_pymorphy, download=True)
sentences = ["Я шёл домой по незнакомой улице.", "Девушка пела в церковном хоре о всех уставших в чужом краю."]
batch_size = 32
for i in range(0, len(sentences), batch_size):
  batch = sentences[i:i+batch_size]
  for parsed_item in model(batch):
    print(parsed_item)
```

File: deeppavlov/models/morpho_tagger/common_tagger.py [hunk 2]
```diff
@@ -2,6 +2,7 @@
 File containing common operation with keras.backend objects
 """
 
+from typing import Union, Optional, Tuple
```
[the remark this answers is not in the document]
Comment (AlexeySorokin) [reply deeppavlov/models/morpho_tagger/common_tagger.py]: Fixed.

File: deeppavlov/models/morpho_tagger/common_tagger.py [hunk 3]
```diff
@@ -25,9 +26,88 @@ def repeat_(x, k):
     return kb.tile(x[:, None, :], tile_factor)
 
 
-def make_pos_and_tag(tag):
-    if "," in tag:
-        pos, tag = tag.split(",", maxsplit=1)
+def make_pos_and_tag(tag: str, sep: str = " ",
```
[the remark this answers is not in the document]
Comment (AlexeySorokin) [reply deeppavlov/models/morpho_tagger/common_tagger.py]: Fixed.

File: deeppavlov/models/morpho_tagger/common_tagger.py [hunk 4]
```diff
@@ -25,9 +26,88 @@ def repeat_(x, k):
...
+        pos, tag = tag.split(sep, maxsplit=1)
     else:
-        pos, tag = tag, "_"
+        pos, tag = tag, ("_" if return_mode is None else "")
+    if return_mode in ["dict", "list", "sorted_dict"]:
+        tag = tag.split("|") if tag != "" else []
+        if "dict" in return_mode:
+            tag = dict(tuple(elem.split("=")) for elem in tag)
+            if return_mode == "sorted_dict":
+                tag = tuple(sorted(tag.items()))
```
[the remark this answers is not in the document]
Comment (AlexeySorokin) [reply deeppavlov/models/morpho_tagger/common_tagger.py]: Yes, I'm sure, but I recognize that the name is confusing. I change it to "sorted_items".

File: deeppavlov/models/morpho_tagger/common_tagger.py [hunk 5]
```diff
@@ -25,9 +26,88 @@ def repeat_(x, k):
...
+    return pos + sep + tag
+
+
+def _are_equal_pos(first, second):
+    NOUNS, VERBS, CONJ = ["NOUN", "PROPN"], ["AUX", "VERB"], ["CCONJ", "SCONJ"]
+    return (first == second or any((first in parts) and (second in parts)
+                                   for parts in [NOUNS, VERBS, CONJ]))
+
+
+IDLE_FEATURES = ["Voice", "Animacy", "Degree", "Mood", "VerbForm"]
```
[the remark this answers is not in the document]
Comment (AlexeySorokin) [reply deeppavlov/models/morpho_tagger/common_tagger.py]: Fixed. But I am not sure sets are faster than lists when they contain only 5 items.

File: deeppavlov/models/morpho_tagger/lemmatizer.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
+import pathlib
+from collections import defaultdict
+import re
+from typing import List, Dict, Generator, Tuple, Any, AnyStr, Union, Optional
+from abc import abstractmethod
+import numpy as np
```
[the remark this answers is not in the document]
Comment (AlexeySorokin) [reply deeppavlov/models/morpho_tagger/lemmatizer.py]: Seemingly fixed.

File: deeppavlov/models/morpho_tagger/lemmatizer.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+from deeppavlov.core.models.serializable import Serializable
+from deeppavlov.core.common.registry import register
+from deeppavlov.models.morpho_tagger.common_tagger import make_pos_and_tag, get_tag_distance
+
+
+class BasicLemmatizer(Serializable):
+    """
+    A basic class for lemmatizers. It must implement two abstract methods:
+    * :meth: `_lemmatize` for single word lemmatization,
+    * :meth: `__call__` for lemmatizing a batch of sentences.
```
[the remark this answers is not in the document]
Comment (AlexeySorokin) [reply deeppavlov/models/morpho_tagger/lemmatizer.py]: Fixed.

File: docs/conf.py [hunk 8]
```diff
@@ -194,7 +194,8 @@
                         'sortedcontainers', 'russian_tagsets', 'bert_dp', 'aiml']
 
 extlinks = {
-    'config': (f'https://github.com/deepmipt/DeepPavlov/blob/{release}/deeppavlov/configs/%s', None)
+    'config': (f'https://github.com/deepmipt/DeepPavlov/blob/{release}/deeppavlov/configs/%s', None),
+    'morpho_config': (f'https://github.com/deepmipt/DeepPavlov/tree/pymorphy_lemmatizer/deeppavlov/configs/%s', None)
```
[the remark this answers is not in the document]
Comment (AlexeySorokin) [reply docs/conf.py]: Fixed.

File: tests/test_quick_start.py [hunk 9]
```diff
@@ -462,7 +464,7 @@ def test_consecutive_training_and_interacting(self, model, conf_file, model_dir,
             logfile = io.BytesIO(b'')
             p = pexpect.popen_spawn.PopenSpawn(sys.executable + " -m deeppavlov train " + str(c), timeout=None,
                                                logfile=logfile)
-            p.readlines()
+            s = p.readlines()
```
[the remark this answers is not in the document]
Comment (AlexeySorokin) [reply tests/test_quick_start.py]: For debugging purposes, fixed.

File: deeppavlov/dataset_iterators/morphotagger_iterator.py [hunk 10]
```diff
@@ -112,9 +110,9 @@ def gen_batches(self, batch_size: int, data_type: str = 'train',
...
             batch_size = L
         for start in range(0, L, batch_size):
-            indexes_to_yield = indexes[start:start+batch_size]
+            indexes_to_yield = indexes[start:start + batch_size]
             data_to_yield = tuple(list(x) for x in zip(*([data[i] for i in indexes_to_yield])))
             if return_indexes:
                 yield indexes_to_yield, data_to_yield
             else:
-                yield data_to_yield
+                yield data_to_yield
```
[the remark this answers is not in the document]
Comment (AlexeySorokin) [reply deeppavlov/dataset_iterators/morphotagger_iterator.py]: Fixed.

File: docs/components/morphotagger.rst (hunk 1 above)
Comment (AlexeySorokin) [reply docs/components/morphotagger.rst]: I do not think it is the work of the user to create batches. This function was one of the most frequently used in my educational work with DeepPavlov, so I find it very convenient.

**Pull request merged**